### PR TITLE
Import/Export

### PR DESCRIPTION
**Pull request by nodar-chkuaselidze**
- Exports Wallet with cosigners and account.
- This format can be reused in bcoin as well (maybe minor improvements) and should be compatible
- :x: Currently does not export/import imported addresses.
- :x: received, change and nested depth is not recovered (for now)

### CODE REVIEW

File: lib/primitives/cosigner.js [hunk 1]
```diff
@@ -235,6 +235,13 @@ class Cosigner extends Struct {
       'tokenDepth must be an u32.');
     this.tokenDepth = json.tokenDepth;
 
+    assert((json.purpose >>> 0) === json.purpose);
+    assert(json.purpose >= 0);
+    this.purpose = json.purpose;
+
+    assert((json.fingerPrint >>> 0) === json.fingerPrint);
```
Comment (tynes) [inline lib/primitives/cosigner.js]: Should the `fingerPrint` be optional for cases in which it cannot be calculated?

Comment (nodar-chkuaselidze) [reply lib/primitives/cosigner.js]: good catch. Actually we are using `-1`, when the information is missing. So I will update to reflect it.

File: lib/export.js [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+
+const NULL_TOKEN = Buffer.alloc(32, 0);
+const NULL_KEY = Buffer.alloc(33, 0);
+
+/**
+ * Cosigner dump object.
+ * Wraps Cosigner primitive.
+ */
+
+class CosignerDetails extends Struct {
```
Comment (tynes) [inline lib/export.js]: Why is there a `CosignerDetails` object when it doesn't add anything additional to the `Cosigner`?

File: lib/export.js [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+  constructor() {
```
Comment (tynes) [inline lib/export.js]: Should the constructor be able to take the `Cosigner` as well? Its not very useful to use `new CosignerDetails()` unless you wanted a nil cosigner object

File: lib/export.js [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+  getJSON(network) {
+    return {
+      watchOnly: this.watchOnly,
+      accountDepth: this.accountDepth,
+      tokenDepth: this.tokenDepth,
+      token: this.token.toString('hex'),
+      master: this.master.toRaw().toString('hex'),
+      joinPubKey: this.joinPubKey.toString('hex'),
+      timestamp: this.timestamp,
+
```
Comment (tynes) [inline lib/export.js]: Nit: empty space

File: lib/export.js [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+    bw.writeU8(flags);
+    bw.writeU32(this.accountDepth);
+    bw.writeBytes(this.token);
+    bw.writeU32(this.tokenDepth);
+    this.master.toWriter(bw);
+
+    bw.writeU40(this.timestamp);
+    bw.writeBytes(this.joinPubKey);
+
+    bw.writeU8(this.accounts.length);
```
Comment (tynes) [inline lib/export.js]: I don't see any checks to make sure this won't overflow

File: lib/multisigdb.js [hunk 6]
```diff
@@ -512,6 +512,132 @@ class MultisigDB extends EventEmitter {
...
+      return await this._export(wid);
+    } finally {
+      unlock();
+    }
+  }
+
+  /**
+   * Export wallet. (without lock)
+   * @param {Number} wid
+   * @returns [Wallet, Account[], Cosigner[]]
```
Comment (tynes) [inline lib/multisigdb.js]: This docstring looks out of date

File: lib/multisigdb.js [hunk 7]
```diff
@@ -512,6 +512,132 @@ class MultisigDB extends EventEmitter {
...
+    assert(Array.isArray(winfo.cosigners));
+    assert(winfo.cosigners.length > 1);
+
+    const author = winfo.cosigners[0].cosigner;
+    const account = winfo.accounts[0];
+
+    assert(account.type === Account.types.MULTISIG);
+    assert(account.name === 'default');
+    assert(account.initialized === true);
+    assert(account.accountIndex === 0);
```
Comment (tynes) [inline lib/multisigdb.js]: Why does the `accountIndex` need to be `0` here?

Comment (nodar-chkuaselidze) [reply lib/multisigdb.js]: Multisig wallets only have one account in them, so it should never be more than 0.

Comment (nodar-chkuaselidze) [reply lib/multisigdb.js]: Multisig wallets only have one account in them, so it should never be more than 0.

File: lib/export.js (hunk 5 above)
Comment (nodar-chkuaselidze) [reply lib/export.js]: Multisig only has one account, but true this needs change to be useful for bcoin.

File: lib/export.js (hunk 3 above)
Comment (nodar-chkuaselidze) [reply lib/export.js]: All these serializers give you default object. So I don't think having many ways to pass cosigner will improve much. (Especially these always clone the data passed.)

File: lib/export.js (hunk 2 above)
Comment (nodar-chkuaselidze) [reply lib/export.js]: To make better wrapper for serializers. Cosigner unlike Wallet and Account is already serializable so I just wrapped it so it does not expose functionality that wont work (mostly client side)

File: lib/multisigdb.js (hunk 7 above)
Comment (tynes) [reply lib/multisigdb.js]: That would be good to add as a comment

**Pull request merged**